06.Lesson_ThinkingProperties/ThinkingPropertiesExercise/PopsicleFinance/README.md

```diff
@@ -0,0 +1,29 @@
+(Did not have as much time to work on and organize this one)
+
+# Categories
+
+## Unit tests
+- ❌ Unit test for `deposit` - check that rewards and balances are updated
+    - ❓ Is there any way to check things about properties of `accounts` without making it public or using a harness? Would this be a good use of a `ghost`?
```

Comment:
Yes you can use a ghost here. You will probably need two, one for each value of the struct. You create two ghost mappings and then two hooks that update the ghosts with the values that are loaded. Here is an example of that being done for a normal private variable: https://github.com/saddle-finance/saddle-contract/blob/hannes/certora/certora/spec/Swap.spec#L127-L136

For this mapping of address to struct you would do something similar except you need the key and an offset to specify which value in the struct to look it. So you would define ghost mappings `accountFees` and `accountRewards` and the hook would read something like: 

```
hook Sload accounts[KEY address account].(offset 0)  uint256 feesCollectedPerShare STORAGE {
    require accountFees[account] == feesCollectedPerShare;
}

hook Sload accounts[KEY address account].(offset 32)  uint256 rewards STORAGE {
    require accountRewards[account] == rewards;
}

Reply:
Looks like you already did this in a more readable way.

Reply:
If I was able to do this _without_ the `offset`, why would we ever want to use the `offset`? In the Saddle example you shared, I see that [here](https://github.com/saddle-finance/saddle-contract/blob/hannes/certora/certora/spec/Swap.spec#L155) they wrote a hook without `offset` but [here](https://github.com/saddle-finance/saddle-contract/blob/hannes/certora/certora/spec/Swap.spec#L166) they used `offset`.  The `offset` seems hard to reason about reliably, so it would be nice if we didn't need it in these cases.

Reply:
I think previously it might've been not an option without using offset. I agree with you though, I don't think there's any use offset anymore.

Reply:
In the saddle case its different than here though its an array in struct while here its a struct in a mapping. To access the struct there we have to use currentContract.<struct>.<structElement>[KEY (optional, if element is an array or mapping)]